### PR TITLE
fix(base): Fix compiler warnings

### LINKED ISSUE #767: Should adapter_, root_ and foundation_ be @protected instead of @private?
According to Closure's [annotating documentation](https://github.com/google/closure-compiler/wiki/Annotating-JavaScript-for-the-Closure-Compiler#protected), protected properties are available to instance methods of any subclass. With `@private`, compilation seems to trigger a warning ("Access to private property adapter_ of XXX not allowed here").

### PR DESCRIPTION
**Pull request by nckh**
Fixes #767 

- Change `@private` to `@protected`
- Fix optional annotations
- Fix enumeration annotations

### CODE REVIEW

File: packages/mdc-base/foundation.js [hunk 1]
```diff
@@ -19,21 +19,21 @@
  */
 export default class MDCFoundation {
 
-  /** @return enum{cssClasses} */
+  /** @return {!Object<string>} */
```
Comment (traviskaufman) [inline packages/mdc-base/foundation.js]: How come these `enum{...}` return types were changed? They seemed to work fine when we tested them.

Comment (nckh) [inline packages/mdc-base/foundation.js]: I've tried multiple combinations and it seems the compiler ignores what follows `@return` when it is not wrapped with curly braces. `@return askjdlajs` passes but `@return {askjdlajs}` triggers the correct warning.

Then I'm not sure whether `@enum` can be used to annotate the return value of a method. This rightfully fails:

```
/**
 * @enum {string}
 */
this.strings = {
  foo: 'bar',
};
``` 

But this does not even though it's super wrong:

```
/**
 * @enum {string}
 */
getStrings() {
  return 123;
}
```

Whereas this throws a warning:

```
/**
 * @return {!Object<string>}
 */
getStrings() {
  return 123;
}
```

I didn't managed to get the object values type checked though. The code below should fail in theory.

```
/**
 * @return {!Object<string>}
 */
getStrings() {
  return {
    foo: 'bar',
    glagla: {},
  };
}
```

**Comment by traviskaufman** [comment]
> Then I'm not sure whether @enum can be used to annotate the return value of a method.

[It can](https://closure-compiler.appspot.com/home#code%3D%252F%252F%2520%253D%253DClosureCompiler%253D%253D%250A%252F%252F%2520%2540compilation_level%2520ADVANCED_OPTIMIZATIONS%250A%252F%252F%2520%2540output_file_name%2520default.js%250A%252F%252F%2520%2540formatting%2520pretty_print%250A%252F%252F%2520%253D%253D%252FClosureCompiler%253D%253D%250A%250A%250A%252F**%2520%2540enum%2520%257Bstring%257D%2520*%252F%250Aconst%2520ALPHABET%2520%253D%2520%257B%250A%2520A%253A%2520'A'%252C%250A%2520B%253A%2520'B'%252C%250A%2520C%253A%2520'C'%252C%250A%257D%253B%250A%250A%252F**%2520%2540return%2520enum%257BALPHABET%257D%2520*%252F%250Afunction%2520getAlphabet()%2520%257B%250A%2520%2520return%2520ALPHABET%253B%250A%257D%250A%250Aconsole.log(getAlphabet().A)%253B%250A%252F%252F%2520Emits%2520warning%253A%250A%252F%252F%2520console.log(getAlphabet().D)%253B). A little trick of the closure compiler :)

**Comment by nckh** [comment]
Oh sweet, thanks! Wish that was documented. Indeed it's checked correctly. Originally my problem was when compiling with `jscomp_warning: 'lintChecks',` I was getting the following warnings:

> ../src/packages/mdc-base/foundation.js:22: WARNING - Missing type declaration.
>   /** @return enum{cssClasses} */
>       ^
> 
> ../src/packages/mdc-base/foundation.js:29: WARNING - Missing type declaration.
>   /** @return enum{strings} */
>       ^
> 
> ../src/packages/mdc-base/foundation.js:36: WARNING - Missing type declaration.
>   /** @return enum{numbers} */
>       ^

But this flag also returns many wrong warnings like missing semicolons so I'm going to turn it off for now.

**Comment by nckh** [comment]
Hi! Do you have any other pending change requests?

**Comment by bogusred** [comment]
Hey Greg. FYI, my [pull request](https://github.com/bogusred/material-components-web/commit/eb98401a0711cba46079a6064bd7f2331ffaa5b5) went through which changes `@private` to `@protected`. Sorry about that! We can chat in our next meeting about other material components to tackle.

**Comment by nckh** [comment]
@bogusred I sync'd with your changes but there are still some private/protected changes and optionals annotation fixes left.

@lynnjepsen please let me know if there's any other change needed. Thanks!

**Pull request merged**